## Medidas de variabilidad

> Las medidas de variabilidad son herramientas estadísticas que nos permiten conocer cuánto se dispersan o alejan los datos respecto a una medida central, como la media. Nos indican qué tan homogéneos o heterogéneos son los datos.

- Datos numéricos
  - La desviación estándar
  - El rango intercuaril
- Datos categóricos
  - Usualmente no se mide la variabilidad

### Desviación estándar

> Mide la dispersión promedio de los datos respecto a la media, en las mismas unidades que los datos

- $s$ desviación estándar de la muestra
- $\sigma$ desviación estándar de la población

$$s=\sqrt{\frac{\sum_{i=1}^{n}(x_i-\overline{x})^2}{n-1}}$$

donde:

- $n$: número de datos que contiene la variable
- $\overline{x}$: promedio de la variable
- $x_i$: cada uno de los datos de la variable
- $n-1$: número de grados de libertad

$$\rho=\sqrt{\frac{\sum_{i=1}^{n}(x_i-\mu)^2}{N}}$$

donde:

- $N$: total de la población
- $\overline{x}$: promedio de la variable
- $x_i$: cada uno de los datos de la variable
- $\mu$: promedio de la población

La desviación estándar es una medida propia de la **distribución normal**, se debería usar si los datos numéricos tienen este tipo de distribución.

In [2]:
import numpy as np
import pandas as pd

In [3]:
datos = np.array([7,5,3,8])
print(f'Media = {np.mean(datos)}, s = {np.std(datos)}')

Media = 5.75, s = 1.920286436967152


Los datos anteriores se alejan 1.92 unidades de la media que es 5.75

In [4]:
datos1 = np.array([4,1,10,8])
datos2 = np.array([4,1,10,8,300])
print('Variable sin valores extremos:')
print(f'Media = {np.mean(datos1)}, s = {np.std(datos1)}')
print('-'*20)
print('Variable que tiene valores extremos:')
print(f'Media = {np.mean(datos2)}, s = {np.std(datos2)}')

Variable sin valores extremos:
Media = 5.75, s = 3.491060010942235
--------------------
Variable que tiene valores extremos:
Media = 64.6, s = 117.74141157638633


Como en la media, la desviación estándar también es **sensible a valores extremos**. Se puede minimizar usando el **rango intercuartiles**, una medida de variabilidad más robusta frente a la presencia de valores extremos.

### Los percentiles

> Son medidas estadísticas que dividen un conjunto de datos ordenado en 100 partes iguales, y cada percentil indica la posición relativa de un valor dentro del conjunto de datos.

#### Ejemplos

Si un estudiante está en el percentil 80 en un examen, significa que superó al 80% de los estudiantes y solo el 20% obtuvo una calificación igual o mayor.

In [5]:
datos = np.array([2,18,8,15,5,13,9,12,11,7])
print(f'Original: {datos}')
print(f'Organizado: {np.sort(datos)}')

Original: [ 2 18  8 15  5 13  9 12 11  7]
Organizado: [ 2  5  7  8  9 11 12 13 15 18]


¿Percentil 10?

- El percentil 10 sería el valor numérico por debajo del cual se encuentra el 10% de los datos y por encima del cual está el 10% de los datos y por encima del cual está el otro 90%
- Si tenemos 10 datos, el 10% de esto será 1 dato. Es decir que el percentil 10 será el valor numérico por debajo del cuál se encuentra 1 dato y por encima del cual se encuentran los 9 datos restantes

In [6]:
#Cálculo del percentil 10
ps = [10]
percentiles = np.percentile(datos, ps)

#Mostrar los percentiles calculados
print(f'Organizado: {np.sort(datos)}')
for p, pv in zip(ps, percentiles):
    print(f'Percentil {p}: {pv}')

Organizado: [ 2  5  7  8  9 11 12 13 15 18]
Percentil 10: 4.7


Para el arreglo anterior el percentil 10 es igual a 4.7. Esto indica:

Debajo del 4.7 tendremos el 10% de los datos. Encima del 4.7 tendremos el 90% de los datos. TEnemos que 4.7 (percentil 10), tendremos el valor 2 (1 dato). Mientras que por encima tendremos los datos 5 7 8 9 11 12 13 15 18 (9 datos).

Tenemos los tres percentiles de interés cuando analizamos la dispersión de los datos:

- Percentil 25: llamado también primer cuartil.
- Percentil 50: llamado también segundo cuartil (que es exactamente igual a la mediana de los datos).
- Percentil 75: llamado también tercer cuartil

Se denominan cuartiles pues dividen la variable en cuatro partes iguales.

In [7]:
ps = [25,50,75]
cuartiles = np.percentile(datos, ps)

print(f'Datos organizados: {np.sort(datos)}')
for p, pv in zip(ps, cuartiles):
    print(f'Percentil {p}: {pv}')

Datos organizados: [ 2  5  7  8  9 11 12 13 15 18]
Percentil 25: 7.25
Percentil 50: 10.0
Percentil 75: 12.75


### Intercuartiles

> El rango intercuartiles (IQR por sus siglas en inglés) de una variable es una medida de dispersión que se calcula como la diferencia entre el cuartil 3 y el cuartil 1 

¿Y por qué es una medida de dispersión? Porque recordemos que entre el cuartil 3 y el 1 está el cuartil 2, que es la misma mediana. Así:

- Si el rango intercuartiles es "grande" quiere decir que los cuartiles 3 y 1 estarán "alejados" de la mediana: "alta" dispersión.
- Si el rango intercuartiles es "pequeño" quiere decir que los cuartiles 3 y 1 estarán "más cerca" de la mediana: "baja" dispersión.

#### Ejemplo

In [8]:
#Calculemos su rango intercuartiles y su mediana
datos = np.linspace(-5,5,10) #datos con 10 valores entre -5 y 5

#Cálculo de cuartiles 1,2 (mediana) y 3:
q1 = np.percentile(datos, 25)
q2 = np.percentile(datos, 50)
q3 = np.percentile(datos, 75)

#Rango intercuartiles
iqr = q3 - q1

#IMprimir información en pantalla
print(f'Datos: {datos}')
print(f'Meidana: {q2}, IQR: {iqr}')

Datos: [-5.         -3.88888889 -2.77777778 -1.66666667 -0.55555556  0.55555556
  1.66666667  2.77777778  3.88888889  5.        ]
Meidana: 0.0, IQR: 5.000000000000001


Ahora con datos más dispersos

In [9]:
#Calculemos su rango intercuartiles y su mediana
datos = np.linspace(-25,25,10) #datos con 10 valores entre -25 y 25

#Cálculo de cuartiles 1,2 (mediana) y 3:
q1 = np.percentile(datos, 25)
q2 = np.percentile(datos, 50)
q3 = np.percentile(datos, 75)

#Rango intercuartiles
iqr = q3 - q1

#IMprimir información en pantalla
print(f'Datos: {datos}')
print(f'Meidana: {q2}, IQR: {iqr}')

Datos: [-25.         -19.44444444 -13.88888889  -8.33333333  -2.77777778
   2.77777778   8.33333333  13.88888889  19.44444444  25.        ]
Meidana: 0.0, IQR: 24.999999999999996


Ambos datos tienen un valor central (mediana) igual a cero. El primero tiene un IQR de 5 mientras que el segundo uno de casi 24.9. El segundo esta más disperso (se alejan más de la mediana) que el primero.

El rango intercuartiles también es menos sensible a la presencia de outliers en comparación con la desviación estándar.

In [10]:
#Calculemos su rango intercuartiles y su mediana
datos = np.linspace(-5,5,10) #datos con 10 valores entre -5 y 5
datos = np.append(datos, [600])

#Cálculo de cuartiles 1,2 (mediana) y 3:
q1 = np.percentile(datos, 25)
q2 = np.percentile(datos, 50)
q3 = np.percentile(datos, 75)

#Rango intercuartiles
iqr = q3 - q1

#IMprimir información en pantalla
print(f'Datos: {datos}')
print(f'Meidana: {q2}, IQR: {iqr}')

Datos: [-5.00000000e+00 -3.88888889e+00 -2.77777778e+00 -1.66666667e+00
 -5.55555556e-01  5.55555556e-01  1.66666667e+00  2.77777778e+00
  3.88888889e+00  5.00000000e+00  6.00000000e+02]
Meidana: 0.5555555555555554, IQR: 5.555555555555556


Cambio en la mediana (0.55 vs. 0) y en el IQR (5.55 vs. 5.0) con valores con y sin valores extremos. El cambio es mínimo a pesar del valor tan alto del outlier.

| | ¿Cuándo usar? | Limitaciones |
|---|---|---|
| Desviación estándar | Distribución normal y no hay presencia de *outliers* | Sensible a los *outliers* |
| Rango intercuartiles | Distribución no normal y cuando hay valores extremos | |

### Ejemplo práctico

In [11]:
df = pd.read_csv('./crimenes.csv')
df

,State,Population,Murder.Rate,Abbreviation
0,Alabama,4779736,5.7,AL
1,Alaska,710231,5.6,AK
2,Arizona,6392017,4.7,AZ
3,Arkansas,2915918,5.6,AR
4,California,37253956,4.4,CA
5,Colorado,5029196,2.8,CO
6,Connecticut,3574097,2.4,CT
7,Delaware,897934,5.8,DE
8,Florida,18801310,5.8,FL
9,Georgia,9687653,5.7,GA


In [12]:
print('Population')
print(f"\tMedia: {df['Population'].mean()}")
print(f"\tDesviacion: {df['Population'].std()}")

print('='*30)
print('Murder.Rate:')
print(f"\tMedia: {df['Murder.Rate'].mean()}")
print(f"\tDesviacion: {df['Murder.Rate'].std()}")

Population
	Media: 6162876.3
	Desviacion: 6848235.347401142
Murder.Rate:
	Media: 4.066
	Desviacion: 1.915736124302923


Rango de intercuartiles 

In [13]:
def calcular_iqr(data):
    #Calcular cuartiles 1 y 3
    qs = data.quantile([0.25, 0.75])

    #Calcular IQR
    iqr = qs[0.75] - qs[0.25]
    return iqr

In [14]:
print('Population')
print(f"\tMediana: {df['Population'].median()}")
print(f"\tIQR: {calcular_iqr(df['Population'])}")

print('='*30)
print('Murder.Rate:')
print(f"\tMediana: {df['Murder.Rate'].median()}")
print(f"\tIQR: {calcular_iqr(df['Murder.Rate'])}")

Population
	Mediana: 4436369.5
	IQR: 4847308.0
Murder.Rate:
	Mediana: 4.0
	IQR: 3.125


#### Resumen

| | Media | Desviación | | Mediana | IQR |
|---|---|---|---|---|---|
| Población | 6'162.876 | 6'848.235 | | 4'436.369 | 4'847.308 |
| Tasa de homicidios | 4.07 | 1.91 | | 4.0 | 3.125 |


En la población hay diferencias importantes entre media/desviación vs. mediana/IQR. Porque probablemente la variable población contiene valores extremos y por lo tanto la mediana y el IQR son más confiables.

En el caso de la tasa de homicidios vemos medidas comparables.

La combinación mediana/IQR es más confiable que la media/desviación.